Adding 'Valid' and 'Same' padding option in Transposed Convolution layer. #2114

kartikdutt18 opened the pull request:
hi everyone,
Added same and valid padding options for transposed convolutions.
Added more tests in ann_layer_test.cpp

zoq commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -407,6 +428,43 @@ void TransposedConvolution<
+    OutputDataType
+>::InitializeSamePadding(){
+  /*
+  * Using O=s*(I-1) + K -2P + A
```
Pedantic style issue, this is missing one space (should align with the first `*` above).

zoq commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -407,6 +428,43 @@ void TransposedConvolution<
+  * P=Padding
+  */
+
+  padWidth = (kernelWidth-strideWidth)/2 + kernelWidth-strideWidth%2 == 1;
```
Another pedantic style issue, missing space between operations (`-,/%`).

zoq commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -407,6 +428,43 @@ void TransposedConvolution<
+
+  padWidth = (kernelWidth-strideWidth)/2 + kernelWidth-strideWidth%2 == 1;
+  padHeight = (kernelHeight-strideHeight)/2 + kernelHeight-strideHeight%2 == 1;
+  // If Padding is negative set it to 0.
```
Hm, the padding can be < 0 if the user passed in valid parameters? If not, instead of setting  `padWidth = 0`, we should print out a failure and exit.

kartikdutt18 replied:
With Same padding if stride width < kernel width then calculated padWidth will be negative which is not possible so if such a case is not possible there is already "Output dimension not possible " log.

zoq replied:
Thanks for the clarification, in this case let's add another fatal error here.

kartikdutt18 replied:
I have added fatal error instead of setting it to 0. Please have a look.

zoq commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -407,6 +428,43 @@ void TransposedConvolution<
+  {
+    padHeight = 0;
+  }
+  aW  = size_t((kernelWidth-strideWidth)%2 == 1 && strideWidth == 1);
```
See comment above.

zoq commented on src/mlpack/tests/ann_layer_test.cpp:
```diff
@@ -1850,6 +1850,31 @@ BOOST_AUTO_TEST_CASE(SimpleTransposedConvolutionLayerTest)
+  module9.Parameters()(8) = 2.0;
+  module9.Reset();
+  module9.Forward(std::move(input), std::move(output));
+  // Value calculated using torch.nn.functional.conv_transpose2d()
```
Missing stop at the end.

kartikdutt18 replied:
Hi @zoq , I am sorry I wasn't able to understand this I assumed it meant a full stop at the end of the comment . Could you have a look at the changes made now?
Thanks.

zoq replied:
Exactly, looks correct now.

zoq commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -77,9 +80,27 @@ TransposedConvolution<
+
+  if (paddingTypeLow == "valid")
+  {
+    // Set Padding to 0
```
Another minor style issue: missing stop at the end here.

walragatver commented on src/mlpack/tests/ann_layer_test.cpp:
```diff
@@ -1848,8 +1848,33 @@ BOOST_AUTO_TEST_CASE(SimpleTransposedConvolutionLayerTest)
+  // Value calculated using torch.nn.functional.conv2d().
   BOOST_REQUIRE_EQUAL(arma::accu(delta), 7732.0);
+
+  TransposedConvolution<> module8(1, 1, 3, 3, 1, 1, 0, 0, 4, 4, 6, 6, "VALID");
```
It looks like you have not tested the `Same` functionality. It would be great to test it as well.

kartikdutt18 replied:
Sorry about that , I have added them now.

zoq replied:
Nice catch!

kartikdutt18 commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -77,9 +80,27 @@ TransposedConvolution<
+      [](unsigned char c){ return std::tolower(c); });
 
+  aW = 0;
+  aH = 0;
```
Hi @walragatver I tested the code without aW and aH initialized to zero and it kept failing. Earlier these variables were directly assigned values but now I am using 
aW = aW + (outputWidth + 2 * padWidth - kernelWidth) % strideWidth 
So I have kept it for now. Please have a look.
If needed I can remove them and add a ternary expression instead to increment by for same padding.Should I do so?
Thanks

walragatver replied:
I will tell you about this tomorrow.

kartikdutt18 commented:
Hi @zoq , @walragatver I have added more tests for same padding type.Could you have a look and what other changes do I need to make?

walragatver commented on src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp:
```diff
@@ -407,6 +426,41 @@ void TransposedConvolution<
+   * K=Kernel Size
+   * P=Padding
+   */
+  size_t totalPadWidth = (kernelWidth - strideWidth);
```
It looks like the `output` would be similar to `input` only if `stride` is equal to one. 
This what actually TensorFlow does. However, we do not implement it in the similar way. 
We implement such that the input is always similar to the output irrespective of `stride`. So it would be great if you could implement it in the same way.

walragatver replied:
For more info you can refer to the discussion in #2119.

kartikdutt18 replied:
Hi @walragatver,
Yes I have implemented it like it is implemented in Tensorflow. I will make the changes to have output shape equal to input shape.

kartikdutt18 replied:
I think I understand what needs to be done, I will commit required changes hopefully by tomorrow evening. Could you have a look then.

kartikdutt18 replied:
I have made the changes such that output size is same as input size  however I am facing some problem while testing it in C++.
I tested the logic in python, 
`https://colab.research.google.com/drive/174HGBANY-cvnapDOeKrAT971rKx9jFYk`
and implemented the same here. I will try to fix it soon.

kartikdutt18 replied:
@walragatver ,@zoq ,I have fixed the error now, I was missing a pair of parenthesis in the new expression.

kartikdutt18 commented:
Hi everyone, I made all required changes (such that output size is same as input size for "same" padding type) and added some test cases for both same and valid type padding.
Please have a look.
Thanks

zoq commented:
Awesome, will take a look shortly.

walragatver commented on src/mlpack/methods/ann/layer/transposed_convolution.hpp:
```diff
@@ -67,25 +67,27 @@ class TransposedConvolution
-                        const size_t outputHeight = 0);
+                        const size_t outputHeight = 0,
+                        const std::string paddingType = "None");
 
```
Can you add a constructor like the following.
https://github.com/mlpack/mlpack/blob/7f520f032e62c73232cfc097e57824088c839c59/src/mlpack/methods/ann/layer/convolution.hpp#L101

kartikdutt18 replied:
Hey @walragatver, I have added the constructor as in convolutional layer.

zoq commented:
Hm, for some reasons some of the ann test fail, let me know if I should take a closer look into the issue.

kartikdutt18 commented:
> Hm, for some reasons some of the ann test fail, let me know if I should take a closer look into the issue.

Hi @zoq, I have fixed one error that I noticed (I hadn't serialized new parameters that I added for new constructor), this fixed most of the error except one i.e. [mat init::](https://dev.azure.com/mlpack/mlpack/_build/results?buildId=254&view=logs&j=1b38ea94-fede-5667-20eb-44d3de1ec085&t=72087b03-c17c-5a26-b713-236d9bb6c096&l=35).
I ran the some code locally on Linux OS and it passed all test cases however it fails the ann_layer test case on OS X on the same machine. Could you have a look and tell me what I have done wrong / missed?
Thanks.

kartikdutt18 commented:
> Hm, for some reasons some of the ann test fail, let me know if I should take a closer look into the issue.
Hi @zoq, It is working fine on both OS that I have installed, Could you help me where I might have made an error?
Thanks

himanshupathak21061998 commented:
Hey, @kartikdutt18 There, is a problem RecurrentNetworkTest in windows build. It is unrelated to your pull request.

zoq commented:
@kartikdutt18, @himanshupathak21061998 is absolutely right it's on my list of things to debug.

kartikdutt18 commented:
Ohh thanks @zoq ,@himanshupathak21061998 I rebased my PR and ran the tests again, every test passed. However some commit changes from a commit by another member are also being shown.
Changes made:

1. Added new constructor to take padding input as a tuple.
2. Added Same and Valid padding option for Transposed Convolution Layer.
3. Added Tests to test both Same and Valid padding as well as for the new Constructor.

Are there any other changes that I have to implement?
Thanks for the help.

walragatver commented on src/mlpack/methods/ann/activation_functions/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ set(SOURCES
   rectifier_function.hpp
   softplus_function.hpp
   swish_function.hpp
+  mish_function.hpp
```
Hey it looks like mish activation function is still here.

kartikdutt18 replied:
Hey @walragatver, I will revert the changes asap. Are the rest of the changes fine?

kartikdutt18 replied:
Thank you @walragatver, this really helped in removing changes from my second PR.
